core/src/main/java/org/infinispan/container/BoundedSegmentedDataContainer.java

```diff
@@ -0,0 +1,255 @@
+package org.infinispan.container;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.PrimitiveIterator;
+import java.util.Spliterator;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+import org.infinispan.commons.util.ComposeArraySpliterator;
+import org.infinispan.commons.util.ConcatIterator;
+import org.infinispan.commons.util.EntrySizeCalculator;
+import org.infinispan.commons.util.IntSet;
+import org.infinispan.container.entries.CacheEntrySizeCalculator;
+import org.infinispan.container.entries.InternalCacheEntry;
+import org.infinispan.container.entries.PrimitiveEntrySizeCalculator;
+import org.infinispan.eviction.EvictionType;
+import org.infinispan.marshall.core.WrappedByteArraySizeCalculator;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.CacheWriter;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.Policy;
+import com.github.benmanes.caffeine.cache.RemovalCause;
+
+/**
+ * Bounded implementation of segmented data container. Bulk operations (iterator|spliterator) that are given segments
+ * use the segments maps directly to only read the given segments (non segment based just read from bounded container).
+ * <p>
+ * Note this implementation supports both temporary non owned segments and not (L1).
+ * @author wburns
+ * @since 9.3
+ */
+public class BoundedSegmentedDataContainer<K, V> extends DefaultSegmentedDataContainer<K, V> {
+   protected final Cache<K, InternalCacheEntry<K, V>> evictionCache;
+   protected final ConcurrentMap<K, InternalCacheEntry<K, V>> entries;
+
+   public BoundedSegmentedDataContainer(int numSegments, long thresholdSize, EvictionType thresholdPolicy) {
+      super(numSegments);
+
+      Caffeine<K, InternalCacheEntry<K, V>> caffeine = caffeineBuilder();
+
+      switch (thresholdPolicy) {
+         case MEMORY:
+            CacheEntrySizeCalculator<K, V> calc = new CacheEntrySizeCalculator<>(new WrappedByteArraySizeCalculator<>(
+                  new PrimitiveEntrySizeCalculator()));
+            caffeine.weigher((k, v) -> (int) calc.calculateSize(k, v)).maximumWeight(thresholdSize);
+            break;
+         case COUNT:
+            caffeine.maximumSize(thresholdSize);
+            break;
+         default:
+            throw new UnsupportedOperationException("Policy not supported: " + thresholdPolicy);
+      }
+      DefaultEvictionListener evictionListener = new DefaultEvictionListener();
+      evictionCache = applyListener(caffeine, evictionListener, new SegmentMapUpdater()).build();
+      entries = evictionCache.asMap();
+   }
+
+   public BoundedSegmentedDataContainer(int numSegments, long thresholdSize,
```

Comment:
Constructor never used

Reply:
This is more for @pferraro if needs it :D